File: problem5/c-solution/p5.c
/*
 * p5.c
 *
 * Problem 5:
 * What is the smallest positive number that is evenly divisible by all of
 * the numbers from 1 to 20?
 */

#include <stdio.h>
#include <stdlib.h>
#define			TEST		10
#define			DEFAULT		20
#define			REDUNDANT	0

void generate_divisors(int *divisors, int bound)
{
	for (int i = 1; i <= bound; i++)
		divisors[i - 1] = i;

	int num, pos_factor;
	for (int i = bound - 1; i >= 0; i--) {
		num = divisors[i];
		if (num == REDUNDANT)
			continue;
		for (int j = i - 1; j >= 0; j--) {
			pos_factor = divisors[j];
			if (pos_factor == REDUNDANT)
				continue;
			if (num % pos_factor == 0)
				divisors[j] = REDUNDANT;
		}
	}
}

long long find_first_divisible_num(int *divisors, int bound)
{
	long long num = (long long) bound;
	int found = 0;
	int divisor;

	while (!found && num++) {
		found = 1;
		for (int i = 0; i < bound; i++) {
			divisor = divisors[i];
			if (divisor == REDUNDANT)
				continue;
			if (num % divisor != 0) {
				found = 0;
				break;
			}
		}
	}

	return num;
}

int main(int argc, char **argv)
{
	int bound;
	(argc == 2) ? (bound = atoi(argv[1])) : (bound = DEFAULT);

	int divisors[bound];
	generate_divisors(divisors, bound);

	long long num = find_first_divisible_num(divisors, bound);

	printf("\nThe smallest positive number that is divisible by all of");
	printf(" the numbers from 1 to %d is: %llu\n\n", bound, num);

	return 0;
}

File: problem5/java-solution/P5.java
/*
 * P5.java
 *
 * Zeke Reyna
 *
 * Problem 5:
 * What is the smallest positive number that is evenly divisible by all of
 * the numbers from 1 to 20?
 */
import java.util.ArrayList;

public class P5
{

        public static ArrayList<Integer> generate_divisors(int bound)
        {
                ArrayList<Integer> retval = new ArrayList<Integer>();

                int num, i;
                boolean redundant;
                for (num = 1; num <= bound; num++) {
                        redundant = false;
                        for (i = num + 1; i <= bound; i++) {
                                if (i % num == 0) {
                                        redundant = true;
                                        break;
                                }
                        }
                        if (!redundant)
                                retval.add(num);
                }

                return retval;
        }

        public static long find_first_divisible_num(ArrayList<Integer> divis, int bound)
        {
                long num = (long) bound;
                boolean found = false;

                for (; !found; num++) {
                        found = true;
                        for (Integer i : divis) {
                                if (num % i != 0) {
                                        found = false;
                                        break;
                                }
                        }
                }
                return --num;
        }

        public static void main(String args[])
        {
                int bound = (args.length == 1) ? Integer.parseInt(args[0]) : 20;

                ArrayList<Integer> divisors = generate_divisors(bound);

                long num = find_first_divisible_num(divisors, bound);

                System.out.print("\nThe smallest positive number that is divisible");
                System.out.print(" by all of the numbers from 1 to "+bound+" is: ");
                System.out.println(num+"\n");
        }
}
